### PR TITLE
Add SynExpr.Dynamic to SyntaxTree.

### PR DESCRIPTION
**Pull request by nojaf**
A first attempt to solve https://github.com/dotnet/fsharp/issues/12755.
This currently moves the logic from the parser to the type checker.

### CODE REVIEW

File: src/Compiler/Checking/CheckExpressions.fs [hunk 1]
```diff
@@ -5692,6 +5692,19 @@ and TcExprThen cenv (overallTy: OverallTy) env tpenv isArg synExpr delayed =
             TcExprThen cenv overallTy env tpenv isArg (SynExpr.LongIdent (isOpt, SynLongIdent([altId], [], [None]), None, mLongId)) delayed
         | _ -> TcLongIdentThen cenv overallTy env tpenv longId delayed
 
+    // x?k<-42
+    | SynExpr.Set(SynExpr.Dynamic(e1, _, e2, _) , rhsExpr, m) ->
+        let e2 =
```
Comment (dsyme) [inline src/Compiler/Checking/CheckExpressions.fs:5697]: We should factor all the branches of this function into separate functions, like I've been doing for TcExpr and TcPat.  I guess you may as well start with this one

File: src/Compiler/Checking/CheckExpressions.fs [hunk 2]
```diff
@@ -5727,6 +5740,21 @@ and TcExprThen cenv (overallTy: OverallTy) env tpenv isArg synExpr delayed =
 
         TcExprThen cenv overallTy env tpenv false func ((DelayedApp (hpa, isInfix, Some func, arg, mFuncAndArg)) :: delayed)
 
+    // e1?e2
+    | SynExpr.Dynamic(e1, mQmark, e2, _) ->
+         let appExpr =
```
Comment (dsyme) [inline src/Compiler/Checking/CheckExpressions.fs:5745]: Likewise here

**Comment by dsyme** [comment]
Looks good

**Comment by nojaf** [comment]
@dsyme this is ready for review.

File: src/Compiler/SyntaxTree/SyntaxTree.fsi [hunk 3]
```diff
@@ -896,6 +896,9 @@ type SynExpr =
     /// Debug points arising from computation expressions
     | DebugPoint of debugPoint: DebugPointAtLeafExpr * isControlFlow: bool * innerExpr: SynExpr
 
+    /// F# syntax: f?x
+    | Dynamic of funcExpr: SynExpr * qmark: range * argExpr: SynExpr * range: range
```
Comment (dsyme) [inline src/Compiler/SyntaxTree/SyntaxTree.fsi:900]: Could you update changelog.md please with information about this.  Also check for any previous changes you've made to public API surface area which should be mentioned in there, thanks

**Review by dsyme — APPROVED** [review]
One comment about updating changelog.md, though that can be done in a different PR as there are probably multiple things needed there

**Comment by nojaf** [comment]
Alright, you mean https://github.com/dotnet/fsharp/blob/main/release-notes.md#fsharp-compiler-service-main by this right?

**Comment by dsyme** [comment]
> 
Alright, you mean https://github.com/dotnet/fsharp/blob/main/release-notes.md#fsharp-compiler-service-main by this right?

Yes, that's right

**Comment by nojaf** [comment]
Gotcha, see https://github.com/dotnet/fsharp/pull/13143.

**Pull request merged**